### ISSUE TITLE
Feature Request - Patcher

### ISSUE THREAD

**Issue by BiRD4** [LMMS/lmms#3194]
A patcher (like the one in FL Studio) would be useful. While it isn't absolutely necessary for creating sounds, it can make it a lot easier to see what plugins are being used and how they are affecting the sound. The patcher should include a dashboard for automation controls to allow for easy access to a variety of settings on the plugins. The implementation of an X/Y pad would be very much appreciated (as both an effect and a controller on the patcher dashboard) because it can help to visualize what is being automated.

**Comment by sksar** [comment]
I too think that this should be implemented and is really a very useful feature.

**Comment by musikBear** [comment]
One of you @sksar @BiRD4  need to **explain** the actual feature, in a concise and detailed way.
'patcher' means nothing to most, and insight in FLs latests ed. is not common.

**Comment by Umcaruje** [comment]
@musikBear said it all, please explain the feature, or this will be closed.

**Comment by BiRD4** [comment]
A patcher is a window that can show instrument/effect chains. An important feature of a patcher is that patcher presets can be saved, similarly to .xpf files. This way, instrument/effect chains can be transferred between project files much more easily. The patcher can also act as an automation control panel, so that the controls for instruments or effects can be accessed easily.

Here's a picture showing a patcher with three tabs and two separate chains:
https://cloud.githubusercontent.com/assets/20910668/21571708/cd41513a-ce9f-11e6-9f08-9762a45162d4.png
Tabs 2 and 3 each have one controller linked to Monstro.

Without color:
https://cloud.githubusercontent.com/assets/20910668/21571536/5db84a7c-ce9e-11e6-8bb6-6bce52895185.png

Here's a page with a picture of FL Studio's patcher (which is much better looking than my drawing):
https://www.image-line.com/support/FLHelp/html/plugins/Patcher.htm

The red lines represent connections between controllers and plugins.

To summarize, a patcher is a window that contains a graphical representation of instrument/effect chains and automation controls and can save presets for later use.

Hopefully, this has illustrated what a patcher is and why it is useful.

**Comment by tresf** [comment]
@BiRD4 github allows pasting images directly into the bug reports.  Please use that technique to protect against [link rot](https://en.wikipedia.org/wiki/Link_rot).

**Comment by BiRD4** [comment]
@tresf I am a relatively new user, so I don't think I understand how to paste an image properly. Opening the image in paint, copying it, and pasting it in the box gives me a link. How do I do this correctly?

**Comment by zonkmachine** [comment]
Just click on a picture in a folder on your computer and drag it over the comment entry window (Where I'm typing this) and drop it there.
Like this:
https://help.github.com/articles/file-attachments-on-issues-and-pull-requests/

**Comment by tresf** [comment]
> [...] pasting it in the box gives me a link. How do I do this correctly?

That's exactly it.  Click "Preview" tab and move it around as needed. 👍

**Comment by tresf** [comment]
Note, the FX Preset feature lives here, may be a dupe https://github.com/LMMS/lmms/issues/1680

**Comment by BiRD4** [comment]
The picture on that page isn't very specific about the Plugin GUI, but if it were structured to look like the picture of a patcher that I posted earlier, it would be about the same. The only thing missing would be automation controls on other tabs of the window.

**Comment by BiRD4** [comment]
Also, I think that the reason that the pictures were not showing up in the window was that I didn't go to the preview tab first. Thanks for the help!

**Comment by BiRD4** [comment]
Actually, upon further inspection, #1680 isn't a dupe. It only contains one part of the suggestion on this page, which is the ability to save effect chains. My idea also throws automation controls and instruments into it.

**Comment by SirBothersome** [comment]
> A patcher?

@BiRD4, like Carla? http://kxstudio.linuxaudio.org/Applications:Carla

**Comment by BiRD4** [comment]
@SirBothersome The patchbay is the main idea I'm going for. The only thing Carla seems to be missing is an automation tab, but of course that isn't so important unless it's being used in a DAW, like what I'm suggesting. Carla is something pretty good to model a patcher after. Just get rid of all the unneeded parts and add an automation tab for use in LMMS's automation tracks.

In general, though, I think that if you want something to look at for reference, FL Studio's patcher is the best option. (Of course, I know that FL Studio's source code isn't available, but I still prefer using its patcher  as a reference over Carla)
